[고현영] offset, limit 위치 변경 및 filtering 기능 추가

hy6652 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x ] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- offset, limit의 위치 수정
- q 객체를 사용하여 filtering 기능 추가

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 입력 데이터 유효성 검사 (예시 - 프론트)
- 모든 정보를 불러온 후 적용됐던 offset, limit의 위치를 데이터베이스를 불러올 때 제한하는 위치에 두었습니다.
- q 객체를 사용하여 filtering을 구현했습니다.

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 사용자가 발생시킬 수 있는 다양한 예외 케이스에 대해 고민하게 되었습니다. (예시)
- 문자열 데이터 입력 시 마지막에 빈 문자열이 포함되어 있는 경우 데이터 처리가 올바르게 되지 않아 유효성 검사에 빈 문자열 제거하는 로직을 추가하였습니다. (예시)

<br />

## :: 기타 질문 및 특이 사항
- 함수 로직이 너무 긴 것 같습니다. 좀 더 효율적인 방법이 없을지 궁금합니다.(예시)

Commit:
changes

Commit:
Merge branch 'main' into feature/hyunyoung-productsViews

Commit:
ADD: filtering 기능 추가 및 offset, limit 설정

- 피드백 받은 내용을 바탕으로 offset과 limit의 위치를 수정했습니다.
- q 객체를 사용하여 필터링 내용을 추가했습니다.
- result 내용 안에 sub description 내용을  추가하고, 바뀐 모델을 바탕으로 feeling 내용을 추가했습니다.

Commit:
ADD: sub description 추가

Commit:
MODIFT: conflict 수정

Commit:
MODIFY: url 수정

hy6652 commented:
conflict 해결했습니다.

lordmyshepherd commented:
code review 내용 공유
```python
from django.views import View
from django.http  import JsonResponse
from django.db.models import Q

from products.models import *

class ProductListView(View):
    def get(self, request):
        category_id   = request.GET.get('category_id', None)
        offset        = int(request.GET.get('offset', 0))
        limit         = int(request.GET.get('limit', 100))
        ingredient_id = request.GET.getlist('ingredient_id', None)
        skintype_id   = request.GET.getlist('skintype', None)
        feeling_id    = request.GET.get('feeling_id', None)
        badge         = request.GET.getlist('badge', None)

        FILTER_SET = {
            "category_id"   : "category__id",
            "badge"         : "badge__in",
            "ingredient_id" : "productingredient__ingredient__id__in",
            "skintype_id"   : "skintypes__skin_type__id__in",
            "feeling_id"    : "productfeelings__feeling__id__in"
        }

        query = { FILTER_SET[key] : value for key, value in request.GET.items()}
        
        products = Product.objects.filter(**query) \
            .select_related("category") \
            .prefetch_related("productingredient_set__ingredient","productfeelings_set__feeling", "productskintype_set__skin_type", "productimage_set")[offset:limit]

        result = [{
            'id'         : product.id,
            'badge'      : product.badge,
            'productName': product.name,
            'size'       : product.size,
            'price'      : product.price,
            'feeling'    : [feeling.feeling.name for feeling in product.productfeelings_set.all()],
            'ingredient' : [item.ingredient.name for item in product.productingredient_set.all()],
            'skin_type'  : [productskintype.skin_type.name for productskintype in product.productskintype_set.all()],
            'url'        : [img.url for img in product.productimage_set.all()],
            'category'   : {
                'categoryId'         : product.category.id,
                'categoryName'       : product.category.category_name,
                'categoryDescription': product.category.main_description,
                'categorySubDescription': product.category.sub_description
            }
        } for product in products]

        return JsonResponse({'result':result}, status=200)

```

Commit:
MODIFY: filtering에 scent 추가

- filtering에서 badge로 사용하던 것을 지우고 scent로 변경했습니다.
- filtering에 사용하는 이름을 직관적으로 알 수 있게 _id로 변경했습니다.

Commit:
Modify : gitignore 수정

Commit:
Modify : gitignore 2차

Commit:
MODIFU: gitigore 수정

hy6652 commented:
- url 받을 때 사용하는 값을 _id로 변경했습니다.
- csv 파일 gitignore에 추가했습니다.

Commit:
MODIFY: conflict 수정

- conflict 수정했습니다
Merge branch 'main' into feature/hyunyoung-productsViews

hy6652 commented:
conflict 해결했습니다.
